examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,606 @@
+      for (PetscInt p = 0; p < num_points; p++) {
+        point_coords[p * dim + 0] = -PetscCosReal((PetscReal)(p + 1) / (PetscReal)(num_points + 1) * PETSC_PI);
+        if (dim > 1) point_coords[p * dim + 1] = -PetscSinReal((PetscReal)(p + 1) / (PetscReal)(num_points + 1) * PETSC_PI);
+        if (dim > 2) point_coords[p * dim + 2] = PetscSinReal((PetscReal)(p + 1) / (PetscReal)(num_points + 1) * PETSC_PI);
+      }
+      PetscCall(DMSwarmRestoreField(dm_swarm, DMSwarmPICField_coor, NULL, NULL, (void **)&point_coords));
+    }
+    PetscCall(DMSwarmMigrate(dm_swarm, PETSC_TRUE));
+
+    // -- Final particle swarm
+    PetscCall(PetscObjectSetName((PetscObject)dm_swarm, "Particle Swarm"));
+    PetscCall(DMViewFromOptions(dm_swarm, NULL, "-dm_swarm_view"));
+  }
+
+  // Set field values on background mesh
+  PetscCall(DMCreateGlobalVector(dm_mesh, &U_mesh));
+  {
+    DMFunc mesh_solution[1] = {EvalU_proj};
+
+    PetscCall(DMProjectFunction(dm_mesh, 0.0, mesh_solution, NULL, INSERT_VALUES, U_mesh));
+  }
+
+  // Visualize background mesh
+  PetscCall(VecViewFromOptions(U_mesh, NULL, "-u_mesh_view"));
+
+  // Interpolate from mesh to points via PETSc
+  {
+    PetscCall(DMSwarmInterpolateFromCellToSwarm_Petsc(dm_swarm, DMSwarmPICField_u, U_mesh));
+    if (!test_mode) PetscCall(DMSwarmViewXDMF(dm_swarm, "swarm_petsc.xmf"));
+    PetscCall(DMSwarmCheckSwarmValues(dm_swarm, DMSwarmPICField_u, tolerance, EvalU_proj));
+  }
+
+  // Interpolate from mesh to points via libCEED
+  {
+    PetscCall(DMSwarmInterpolateFromCellToSwarm_Ceed(dm_swarm, DMSwarmPICField_u, U_mesh));
+    if (!test_mode) PetscCall(DMSwarmViewXDMF(dm_swarm, "swarm_ceed.xmf"));
+    PetscCall(DMSwarmCheckSwarmValues(dm_swarm, DMSwarmPICField_u, tolerance, EvalU_proj));
+  }
+
+  // Cleanup
+  CeedDestroy(&swarm_ceed_context->ceed);
+  CeedElemRestrictionDestroy(&swarm_ceed_context->restriction);
+  CeedBasisDestroy(&swarm_ceed_context->basis);
+  PetscCall(PetscFree(swarm_ceed_context));
+  PetscCall(DMDestroy(&dm_swarm));
+  PetscCall(DMDestroy(&dm_mesh));
+  PetscCall(VecDestroy(&U_mesh));
+  return PetscFinalize();
+}
+
+// Solution functions
+PetscScalar EvalU(PetscInt dim, const PetscScalar x[]) {
+  PetscScalar result = 1, center = 0.1;
+
+  for (PetscInt d = 0; d < dim; d++) {
+    result *= tanh(x[d] - center);
+    center += 0.1;
+  }
+  return result;
+}
+
+PetscErrorCode EvalU_proj(PetscInt dim, PetscReal t, const PetscReal x[], PetscInt num_comp, PetscScalar *u, void *ctx) {
+  PetscFunctionBeginUser;
+  for (PetscInt c = 0; c < num_comp; c++) u[c] = (c + 1.0) * EvalU(dim, x);
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/*@C
+  DMSwarmCreateReferenceCoordinates - Compute the cell reference coordinates for local DMSwarm points.
+
+  Collective
+
+  Input Parameters:
++ dm_swarm  - the `DMSwarm`
+
+  Output Parameter:
++ is_points    - The IS object for indexing into points per cell
++ X_points_ref - Vec holding the cell reference coordinates for local DMSwarm points
```

Comment:
```suggestion
. dm_swarm  - the `DMSwarm`

  Output Parameter:
+ is_points    - The IS object for indexing into points per cell
- X_points_ref - Vec holding the cell reference coordinates for local DMSwarm points
```
Minor: fix PETSc docstring parameter prefixes

Reply:
I meant to ask about that - what's the magic decoder ring for what the . vs + vs - mean?

Reply:
In PETSc? That's a sowing idiosyncrasy.
https://wgropp.cs.illinois.edu/projects/software/sowing/doctext/node12.htm#Node12